dask_match/core.py

```diff
@@ -616,7 +625,245 @@ def optimize_expr(expr):
             expr = replace_all(expr, replacement_rules)
     finally:
         _defer_to_matchpy = False
+
+    if fuse:
+        # Apply fusion
+        expr = _blockwise_fusion(expr)
+
     return expr
 
 
 from dask_match.reductions import Count, Max, Min, Mode, Size, Sum
+
+
+## Utilites for Expr fusion
+
+
+def replace_nodes(expr, nodes_to_replace: dict):
```

Comment:
I propose making this a method on expr, and naming it `substitute`. 

You might find it interesting to look at sympy's API for the same operation: https://docs.sympy.org/latest/tutorials/intro-tutorial/basic_operations.html#substitution

Reply:
Agree that this logic belongs in `Expr` - I moved over the existing API, but will probably try to make things look more like the sympy API.